// type-predicates-generator/dist/src/cli.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const commander_1 = require("commander");
const generate_1 = require("./generate");
const program = new commander_1.Command();
program
    .option("-p, --project <type>", "Path for project tsconfig.json", "tsconfig.json")
    .option("-f, --file-glob <type>", "file glob pattern target types", "**/*.ts")
    .option("-o, --output <type>", "output file path", "type-predicates.ts")
    .option("-b, --base-path <type>", "project base path", "./")
    .option("-a, --asserts", "generate assert functions or not", false)
    .option("-w, --watch", "watch or not", false)
    .option("--default-array-check-option", "how to check child element type. 'all' or 'first'", "all")
    .option("-c, --comment", "generate JSDoc comments or not", false)
    .option("--whitelist", "not allow non listed properties")
    .parse(process.argv);
const option = program.opts();
const cwd = process.cwd();
(0, generate_1.run)({
    tsconfigPath: (0, path_1.resolve)(cwd, option.project),
    fileGlobs: [option.fileGlob],
    output: (0, path_1.resolve)(cwd, option.output),
    basePath: (0, path_1.resolve)(cwd, option.basePath),
    option: {
        asserts: option.asserts,
        watch: option.watch,
        defaultArrayCheckOption: option.defaultArrayCheckOption,
        comment: option.comment,
        whitelist: option.whitelist,
    },
});
