// type-predicates-generator/dist/src/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.switchExpression = exports.ng = exports.ok = exports.isNg = exports.isOk = void 0;
function isOk(result) {
    return result.__type === "ok";
}
exports.isOk = isOk;
function isNg(result) {
    return result.__type === "ng";
}
exports.isNg = isNg;
function ok(value) {
    return {
        __type: "ok",
        ok: value,
    };
}
exports.ok = ok;
function ng(value) {
    return {
        __type: "ng",
        ng: value,
    };
}
exports.ng = ng;
const toResult = (target, isParentMatch, resolveParent, parentResolved) => {
    const resolved = typeof parentResolved === "undefined"
        ? isParentMatch(target)
            ? resolveParent(target)
            : undefined
        : parentResolved;
    return {
        resolved,
        default: (resolveDefault) => resolved ?? resolveDefault(target),
        case: (isMatch, resolve) => toResult(target, isMatch, resolve, resolved),
    };
};
const switchExpression = (target) => {
    return {
        resolved: undefined,
        default: (resolveDefault) => resolveDefault(target),
        case: (isMatch, resolve) => toResult(target, isMatch, resolve, undefined),
    };
};
exports.switchExpression = switchExpression;
